WIP BXMSDOC-7998: Migration doc for Kogito

hmanwani-rh opened the pull request:
Added "Getting started with migration to Red Hat build of Kogito microservices" doc

See JIRA: https://issues.redhat.com/browse/BXMSDOC-7998

See the following BAPLs:
https://issues.redhat.com/browse/BAPL-1979
https://issues.redhat.com/browse/BAPL-1978
https://issues.redhat.com/browse/BAPL-1977

Doc preview:
[Getting started with migration to Red Hat build of Kogito microservices](http://file.pnq.redhat.com/~hmanwani/BXMSDOC-7998-RHPAM/#assembly-getting-started-migration-to-kogito-microservices)

tarilabs commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-dmn-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,63 @@
+= Migration of a DMN service to {KOGITO} microservice
+
+[role="_abstract"]
+You can migrate decision services that are implemented using DMN models to {KOGITO} microservices by moving the DMN resources from KJAR artifacts to the respective {KOGITO} archetype.
```
I can understand styleguide applies, but from the original draft :

> (...) DMN-based knowledge artifacts can be migrated to Kogito-based deployment with little effort; some additional and idiomatic features of the Kie v7 platform will no longer be required (...)

I would like to understand if there is a way we can indicate:

- migration for DMN-based should be considered _easy_
- some idiomatic features of Kie v7 platform _are no longer required_

hmanwani-rh replied:
Done

tarilabs commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-dmn-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,63 @@
+
+|Object models (POJOs) required for {KIE_SERVER} generic marshalling
+|managed using Data Model object editor in {CENTRAL}
+|object model editing is not supported
```
ref above, I understand styleguide applies, but if compared to provided draft, I would like to understand if there is a way we can indicate:

- this was an idiomatic features of Kie v7 platform and should be _no longer required_

hmanwani-rh replied:
Done

tarilabs commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-dmn-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,63 @@
+
+|{KIE_SERVER} Client API
+|used in conjunction with object models to interact with a KJAR that is deployed on {KIE_SERVER}
+|as for object models, this feature is not supported
```
ref above, I understand styleguide applies, but if compared to provided draft, I would like to understand if there is a way we can indicate:

- this was an idiomatic features of Kie v7 platform and should be _no longer required_ and _Users can choose their own choice of REST library_

hmanwani-rh replied:
Done

tarilabs commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-dmn-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,63 @@
+== Migration strategy
+
+[role="_abstract"]
+When migrating a DMN project to {KOGITO} project, you must first migrate external applications that are interacting with decision services on the {KIE_SERVER} using REST endpoints that are specific to DMN models. For more information about REST endpoints for specific DMN models, see {URL_DEVELOPING_DECISION_SERVICES}#ref-rest-api-dmn-endpoints_dmn-models[_REST endpoints for specific DMN models_].
+
```
This phrase is not entirely semantically equivalent to what originally provided in draft.
As the bulletpoints later below clarifies, we should suggest user to migrate external applications in 2 steps

Step1: migrate external app `from generic {KIE_SERVER} API to specific REST endpoint` (while remaining on KIE_SERVER)
Migration of DMN from kie-server to Kogito
Step2: migrate external app `from {KIE_SERVER} REST endpoint specific to DMN modl to {KOGITO} deployment` 

Here is my proposal of correction:
```suggestion
When migrating a DMN project to {KOGITO} project, you should first migrate external applications that are interacting with decision services on the {KIE_SERVER}, to use {KIE_SERVER} REST endpoints that are specific to DMN models; after this step, it will be easier to migrate the external applications from {KIE_SERVER} to {KOGITO} deployment. For more information about REST endpoints for specific DMN models, see {URL_DEVELOPING_DECISION_SERVICES}#ref-rest-api-dmn-endpoints_dmn-models[_REST endpoints for specific DMN models_].

```

hmanwani-rh replied:
@tarilabs as per the IBM style guide, we need to avoid the usage of `should`, and `should` must be replaced with `must`. Also to make the paragraph easy to understand, rewriting the paragraph as follows:

"When migrating a DMN project to {KOGITO} project, you must first migrate external applications that are interacting with decision services on the {KIE_SERVER}. You can use the REST endpoints that are specific to DMN models. After using the REST endpoints, it is easier to migrate the external applications from {KIE_SERVER} to {KOGITO} deployment. For more information about REST endpoints for specific DMN models, see {URL_DEVELOPING_DECISION_SERVICES}#ref-rest-api-dmn-endpoints_dmn-models[_REST endpoints for specific DMN models_]."

tarilabs commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-external-application-rest.adoc:
```diff
@@ -0,0 +1,15 @@
+= Migrating external applications to DMN model specific REST endpoints
+
+[role="_abstract"]
+To migrate a DMN project to {KOGITO} deployment, you first need to migrate external applications using the REST endpoints that are specific to DMN models. The external applications interact with the decision services on the {KIE_SERVER}.
```
as above:

```suggestion
To migrate a DMN project to {KOGITO} deployment, you first should migrate external applications that are interacting with decision services on the {KIE_SERVER}, to use {KIE_SERVER} REST endpoints that are specific to DMN models. The external applications interact with the decision services on the {KIE_SERVER}.
```

hmanwani-rh replied:
Similarly, changing this paragraph as follows:

To migrate a DMN project to {KOGITO} deployment, you first must migrate external applications that are interacting with decision services on the {KIE_SERVER}, using {KIE_SERVER} REST endpoints that are specific to DMN models.

tarilabs commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-dmn-model-kjar.adoc:
```diff
@@ -0,0 +1,35 @@
+You can use the provided implementation details when your external application is re-routed to the new URL.
+--
+
+After migrating a DMN model KJAR to a {KOGITO} microservice, you need to deploy the microservice using {OPENSHIFT}. For deployment options with Openshift, see {URL_GETTING_STARTED_KOGITO_MICROSERVICES}#con-kogito-operator-deployment-options_deploying-kogito-microservices-on-openshift[_OpenShift deployment options with the RHPAM Kogito Operator_].
```
👍  super, thank you

tarilabs commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-external-application-rest.adoc:
```diff
@@ -0,0 +1,15 @@
+[id="proc-kogito-migrate-external-application-REST_{context}"]
```
In the original draft, there was a "running example":

- showing high level the KJAR content
- showing high level the use of the Kie Server {KIE_SERVER} Client Java API
- showing model-specific {KIE_SERVER} endpoints

and later further referenced in the other .adoc for the migration to Kogito.

Is there a way we can show the content provided in the original draft?

hmanwani-rh replied:
Done, added example of migrating DMN model KJAR

tarilabs reviewed (commented):
Thank you @hmanwani-rh; as agreed, I am highlighting both specific and high level comments as a PR review.

mariofusco commented on doc-content/enterprise-only/getting-started-kogito/con-migration-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,9 @@
+
+The decision services that are implemented using DMN models in {CENTRAL} can be migrated by moving the DMN resources from KJAR artifacts to {KOGITO} archetypes. Similarly, you can migrate the prediction services that are implemented using PMML models by moving the PMML resources from KJAR artifacts to {KOGITO} archetype.
+
+Also, the decision services that are implemented using DRL can be migrated to {KOGITO} deployments using {QUARKUS} REST endpoint. Migrating the DRL files using {QUARKUS} enables you to leverage the {QUARKUS} features while embracing the programming model of {KOGITO}, which is based on the rule units.
```
I'd replace this sentence with something like this

`Also, the decision services that are implemented using DRL can be simply wrapped with {QUARKUS} REST endpoint. However bringing them to {KOGITO} enables you to leverage the main {QUARKUS} features like hot reload and native compilation, while fully embracing the programming model of {KOGITO}, which is based on the rule units, will also enable the automatic generation of those REST endpoints.`

That said I also find that translating the `{QUARKUS}` placeholder with `Red Hat build of Quarkus` (and same thing for Kogito) makes sentences like this extremely hard to read.

hmanwani-rh replied:
Done, changing the paragraph as follows:

Also, the decision services that are implemented using DRL can be enclosed with {QUARKUS} REST endpoint. However, migrating the same decision services to {KOGITO} enables you to use major Quarkus features, such as hot reload and native compilation. Using major Quarkus features and embracing the programming model of Kogito, enables the automatic generation of the Quarkus REST endpoints.

mariofusco commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-drl-expose-ruleunits.adoc:
```diff
@@ -0,0 +1,131 @@
+[id="proc-kogito-migrate-drl-expose-ruleunits_{context}"]
+= Exposing rule units with a REST endpoint using {QUARKUS}
```
There isn't any rule unit up to this point, they will be introduced only with the last step. Please rephrase this as

`Exposing rules evaluation with a REST endpoint using {QUARKUS}`

hmanwani-rh replied:
Done

mariofusco commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-drl-expose-ruleunits.adoc:
```diff
@@ -0,0 +1,131 @@
+[id="proc-kogito-migrate-drl-expose-ruleunits_{context}"]
+= Exposing rule units with a REST endpoint using {QUARKUS}
+
+You can get started with migrating DRL project to {KOGITO} by using REST endpoint support and {QUARKUS}.
```
I wouldn't mention Kogito here, since we are not using it yet in this step. Please rephrase this as

`You can get started exposing the former rules evaluation with a REST endpoint using {QUARKUS}.`

hmanwani-rh replied:
Done

mariofusco commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-drl-implement-ruleunits.adoc:
```diff
@@ -0,0 +1,152 @@
+
+In {KOGITO}, a rule unit contains a set of rules and the facts, against which the rules are matched. Rule units in {KOGITO} also come with data sources. A rule unit data source is a source of the data processed by a given rule unit and represents the entry point, which is used to evaluate the rule unit. There are two types of data sources:
+
+* `DataStream`: This is an appended data source. In `DataStream`, subscribers receive new and past messages, stream can be hot or cold in the reactive streams. Also, `DataStream` cannot be updated or removed.
```
This description is quite misleading. Please replace it with something more simple like

`DataStream: This is an append-only data source, so facts added into DataStream cannot be updated or removed.`

hmanwani-rh replied:
Done

mariofusco commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-drl-implement-ruleunits.adoc:
```diff
@@ -0,0 +1,152 @@
+In {KOGITO}, a rule unit contains a set of rules and the facts, against which the rules are matched. Rule units in {KOGITO} also come with data sources. A rule unit data source is a source of the data processed by a given rule unit and represents the entry point, which is used to evaluate the rule unit. There are two types of data sources:
+
+* `DataStream`: This is an appended data source. In `DataStream`, subscribers receive new and past messages, stream can be hot or cold in the reactive streams. Also, `DataStream` cannot be updated or removed.
+* `DataStore`: This data source is for modifiable data, in which subscribers can access the data store by using the added fact.
```
Also rewrite this as it follows

`DataStore: This data source is for modifiable data. It is possible to update or remove an object using the FactHandle returned when the object is added into the DataStore.`

hmanwani-rh replied:
Done

hmanwani-rh commented:
@mariofusco @tarilabs I've incorporated the suggestions and missing content. Please take a look and share your comments. Also, the doc preview is updated.

tarilabs reviewed (approved):
Thank you for incorporating the suggested changes 👍 
I still have a doubt about the "IBM style guide" but I will follow up on this offline.

gitgabrio reviewed (approved):
Hi @hmanwani-rh Thanks for the PR.
LGTM

sterobin commented on assemblies/assembly-getting-started-migration-to-kogito-microservices.adoc:
```diff
@@ -0,0 +1,40 @@
+[id='assembly-getting-started-migration-to-kogito-microservices']
+= Getting started with migration to {KOGITO} microservices
```
Any reason this is "Getting started with migration ..."? I'd just say "Migrating to ...".

hmanwani-rh replied:
I thought of being consistent with other Getting started book, but I think changing it to "Migrating to..." makes more sense. thank you :)

sterobin commented on assemblies/assembly-getting-started-migration-to-kogito-microservices.adoc:
```diff
@@ -0,0 +1,40 @@
+:context: getting-started-migration-kogito-microservices
+
+// Purpose statement for the assembly
+As a developer of business decisions and processes, you can migrate your decision services that are implemented in {PRODUCT} to {KOGITO} microservices to mold a set of business decisions to your own domain-specific cloud-native set of services.
```
Suggestion to streamline, and for consistency with other intros:

>As a developer of business decisions and processes, you can migrate your decision services in {PRODUCT} to {KOGITO} microservices. When you perform this migration, your existing business decisions become part of your own domain-specific cloud-native set of services. You can migrate Decision Model and Notation (DMN) models, Predictive Model Markup Language (PMML) models, or Drools Rule Language (DRL) rules.

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migration-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,9 @@
+
+The decision services that are implemented using DMN models in {CENTRAL} can be migrated by moving the DMN resources from KJAR artifacts to {KOGITO} archetypes. Similarly, you can migrate the prediction services that are implemented using PMML models by moving the PMML resources from KJAR artifacts to {KOGITO} archetype.
+
+Also, the decision services that are implemented using DRL can be enclosed with {QUARKUS} REST endpoint. However, migrating the same decision services to {KOGITO} enables you to use major Quarkus features, such as hot reload and native compilation. Using major Quarkus features and embracing the programming model of Kogito, also enables the automatic generation of the Quarkus REST endpoints.
```
Suggestion to streamline and clarify, and also to adhere to the define-upon-first-mention rule for modules. I've pulled content from your intros to each of the chapters to seam things together a bit, but I'm not totally clear on the DRL description bullet and intro. So have a look and see if that's correct or needs some clarification. Not super clear to me what precisely the user is to do. 

```
You can migrate the decision service artifacts that you developed in {CENTRAL} to {KOGITO} microservices. {KOGITO} currently supports migration for the following types of decision services:

* *Decision Model and Notation (DMN) models*: You migrate DMN-based decision services by moving the DMN resources from KJAR artifacts to the respective {KOGITO} archetypes.

* *Predictive Model Markup Language (PMML) models*: You migrate PMML-based prediction and decision services by moving the PMML resources from KJAR artifacts to the respective {KOGITO} archetypes.

* *Drools Rule Language (DRL) rules*: You migrate DRL-based decision services by building and deploying a sample project in {KOGITO} to expose a stateless rules evaluation of the decision engine in a {QUARKUS} REST endpoint, and then migrate the REST endpoint to {KOGITO}.

When you migrate your decision services from {CENTRAL} to {KOGITO}, you can also begin using major Quarkus features, such as hot reload and native compilation. The Quarkus features and the programming model of {KOGITO} enables the automatic generation of the Quarkus REST endpoints for implementation in your applications and services.
```

gitgabrio replied:
hi @sterobin @hmanwani-rh 
IMO everything related to PMML should be called "prediction/prediction service", while "decision/decision service" should be used for DMN/DRL.
("You migrate PMML-based prediction and decision services"  -> "You migrate PMML-based prediction and prediction services")
PMML predictions and DMN/DRL decisions are semantically pretty different, I think we should keep clear that.

hmanwani-rh replied:
Done. Changed as following:

```
You can migrate the decision service artifacts that you developed in {CENTRAL} to {KOGITO} microservices. {KOGITO} currently supports migration for the following types of decision services:

* *Decision Model and Notation (DMN) models*: You migrate DMN-based decision services by moving the DMN resources from KJAR artifacts to the respective {KOGITO} archetypes.

* *Predictive Model Markup Language (PMML) models*: You migrate PMML-based prediction and prediction services by moving the PMML resources from KJAR artifacts to the respective {KOGITO} archetypes.

* *Drools Rule Language (DRL) rules*: You migrate the DRL-based decision services by enclosing them in a {QUARKUS} REST endpoint. This approach of migration enables you to use major Quarkus features, such as hot reload and native compilation. The Quarkus features and the programming model of {KOGITO} enables the automatic generation of the Quarkus REST endpoints for implementation in your applications and services.
```

gitgabrio replied:
Thx! :smile:

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-dmn-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,65 @@
+= Migration of a DMN service to {KOGITO} microservice
+
+[role="_abstract"]
+You can migrate decision services that are implemented using DMN models to {KOGITO} microservices easily by moving the DMN resources from KJAR artifacts to the respective {KOGITO} archetype. In the {KOGITO} microservices, some of the KIE v7 features are no longer required.
```
Delete "easily". Per IBM, avoid things like "simply", "easily", etc.

Also, I suggest making this first sentence be whatever you decide on in the bulleted approach from my previous comment. As of now, I'd streamline it a bit like this for consistency and to trim up a bit.

>You can migrate DMN-based decision services by moving the DMN resources from KJAR artifacts to the respective {KOGITO} archetypes.

Also a question: Sometimes you say "archetypes" plural and sometimes singular. I assume we mean " to the respective {KOGITO} archetypes" plural, correct?

hmanwani-rh replied:
Here, the archetype is referring to the respective Kogito archetype that is related to each KJAR artifact.

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-dmn-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,65 @@
+[id="ref-migrate-dmn-to-kogito-considerations_{context}"]
+== Major changes and migration considerations
+
+The following table describes the major changes and features that affect migration from KIE v7 API and KJAR to {KOGITO} deployments:
```
I assume "from **the** KIE v7 API and KJAR", or maybe even "KJARs"

hmanwani-rh replied:
changed it to the {KIE_SERVER} API. 
KJAR is used singular

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-dmn-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,65 @@
+
+The following table describes the major changes and features that affect migration from KIE v7 API and KJAR to {KOGITO} deployments:
+
+.Migration to DMN considerations
```
Suggest "DMN migration considerations", which I think is clearer and more accurate.

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-dmn-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,65 @@
+|In {KIE_SERVER} API
+|In {KOGITO} artifact
+
+|DMN Models
```
Should be lowercase "DMN models" right?

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-dmn-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,65 @@
+
+
+[id="ref-migrate-dmn-to-kogito-considerations_{context}"]
+== Major changes and migration considerations
```
For the table in this section and all other tables in all other sections:
* Suggest capitalizing all sentences starts, as in "Stored in ...", "Copy ...", "Managed using ..." instead of "stored in ...", etc. Not sure of the rationale, and more correct/common to cap all.
* Also per IBM, suggest end all with a period. In tables and lists, if one of the close components is a complete sentence requiring a period, then all other nearby entries should likewise carry a period for consistency, even when not complete sentences. So would be "Stored in src/... of KJAR." and "Copy as is to src/main/resources." etc.

hmanwani-rh replied:
_Suggest capitalizing all sentences starts, as in "Stored in ...", "Copy ...", "Managed using ..." instead of "stored in ...", etc. Not sure of the rationale, and more correct/common to cap all._
I would keep it lower case, as the sentences in the table is a continuation to the header, for example:

In {KIE_SERVER} API, stored in `src/main/resources` of KJAR. 

_Also per IBM, suggest end all with a period. In tables and lists, if one of the close components is a complete sentence requiring a period, then all other nearby entries should likewise carry a period for consistency, even when not complete sentences. So would be "Stored in src/... of KJAR." and "Copy as is to src/main/resources." etc._

agree, adding periods to all sentences.

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-dmn-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,65 @@
+
+|DMN Models
+|stored in `src/main/resources` of KJAR
+|copy as-is to `src/main/resources`
```
Should be "as is", so "Copy as is to ..."  Only hyphenated when adjective, as in "The sale is an as-is agreement."

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-dmn-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,65 @@
+|must be configured using CDI, by annotating the `DMNRuntimeEventListener` with CDI’s `@ApplicationScope` annotation
+
+|Other configuration options
+|configured using system property or `kmodule.xml`
```
"or `kmodule.xml` file"  Per IBM, try to always qualify the mentioned artifact. Ditto everywhere this appears.

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-dmn-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,65 @@
+|{KIE_SERVER} Client API
+|used in conjunction with object models to interact with a KJAR that is deployed on {KIE_SERVER}
+a|as for object models, this feature is no longer required.
+NOTE: You can select your own choice of REST library.
```
Not sure what you mean by "as for object models" Do you mean just "For object models, this ..." Also, the NOTE isn't rendering properly. Seems like it should be since you have the `a|` notation, but check it out. Also look at some other tables with notes in the source to see. I know I did several for the Kogito process BPMN reference content tables. Don't remember off the top of my head, but they worked there.

hmanwani-rh replied:
Done, fixed it

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-dmn-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,65 @@
+
+
+|REST API
+|when KJAR is deployed on {KIE_SERVER}, the applications that interact with an endpoint specific to a DMN model, uses the same API on {KOGITO} deployment
```
"When **a** KJAR is ...", and do you mean "the applications interact ... a DMN model and use the same ..." Otherwise I'm a bit lost in the sentence.

hmanwani-rh replied:
Changed it to: 
`when a KJAR is deployed on {KIE_SERVER}, the applications interacting with specific DMN model endpoint, use the same API on {KOGITO} deployment.`

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-dmn-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,65 @@
+== Migration strategy
+
+[role="_abstract"]
+When migrating a DMN project to {KOGITO} project, first you can migrate external applications that are interacting with decision services on the {KIE_SERVER}. You can use the REST endpoints that are specific to DMN models. After using the REST endpoints, it is easier to migrate the external applications from {KIE_SERVER} to {KOGITO} deployment. For more information about REST endpoints for specif ...
```
"to **a** {KOGITO} project"  I'm noticing articles are being dropped a lot throughout. Just a caution about this I suppose.

Also, this sentence has what's called a dangling modifier:

>After using the REST endpoints, it is easier to migrate the external applications from {KIE_SERVER} to {KOGITO} deployment.

Meaning, "After using the REST endpoints, " what follows must be the subject that's using the endpoints, which is "you". Also, IBM says avoid the empty "it is", "there is/are" , constructions anyway.

So suggest this:

>After using the REST endpoints, you can migrate the external applications from {KIE_SERVER} to {KOGITO} deployment.

If you want to say that something "is easier", that's fine, but need to say easier than what: "... you can more easily migrate the .... than doing ...." or whatnot.

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-dmn-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,65 @@
+[role="_abstract"]
+When migrating a DMN project to {KOGITO} project, first you can migrate external applications that are interacting with decision services on the {KIE_SERVER}. You can use the REST endpoints that are specific to DMN models. After using the REST endpoints, it is easier to migrate the external applications from {KIE_SERVER} to {KOGITO} deployment. For more information about REST endpoints for specif ...
+
+The migration strategy of migrating DMN project to {KOGITO} deployment, includes the following steps:
```
Again I'm getting worried about all these dropped articles that are making the sentences incorrect and telegraphic. Should be "migrating a DMN project to a {KOGITO} deployment". In some cases I think what you're going for is something like, "When you're migrating from project to deployment, ..." like saying, "When you're working week to week", but that's not super clear at first, and in most cases it should have articles anyway. Just my 2 cents.

Also, I'm seeing quote a few unexpected commas here and there throughout. I've recast several sentences previous to this point that eliminated some of the others, but here again should be "to a {KOGITO} project includes ..." Just another caution to be wary of unnecessary commas.

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-dmn-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,65 @@
+
+The migration strategy of migrating DMN project to {KOGITO} deployment, includes the following steps:
+
+. Migrate existing external applications from generic {KIE_SERVER} API to specific REST endpoint for DMN models using the {KIE_SERVER}.
```
"from **the** generic ... to **a** specific"
"to **a/the** {KOGITO} deployment"
"Deploy **a** {KOGITO} microservice"
"from **the** REST endpoint specific to **the** DMN model to **the** {KOGITO} deployment"

hmanwani-rh replied:
Done, fixed it

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-dmn-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,65 @@
+[id="con-migrate-dmn-to-kogito-overview_{context}"]
+= Migration of a DMN service to {KOGITO} microservice
```
"to **a** {KOGITO} microservice"

Also, I really like how you've stuck to the FCC template here! Converting "Migrating" to a noun phrase like this. So many would just keep it a gerund, but well done!

sterobin commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-external-application-rest.adoc:
```diff
@@ -0,0 +1,15 @@
+[id="proc-kogito-migrate-external-application-REST_{context}"]
+= Migrating external applications to DMN model specific REST endpoints
```
Unclear modifier. Suggest " ... to DMN-specific REST endpoints" or "... to REST endpoints specific to DMN models" or even just "... to DMN REST endpoints"

sterobin commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-external-application-rest.adoc:
```diff
@@ -0,0 +1,15 @@
+= Migrating external applications to DMN model specific REST endpoints
+
+[role="_abstract"]
+To migrate a DMN project to {KOGITO} deployment, first you can migrate external applications that are interacting with decision services on the {KIE_SERVER}, using {KIE_SERVER} REST endpoints that are specific to DMN models.
```
"to **a** {KOGITO} deployment"  Also suggest

>... , first you must migrate external applications that use DMN REST endpoints to interact with decision services on the {KIE_SERVER}."

hmanwani-rh replied:
added the article. However, regarding using `must`, SME has specifically asked to change it to can, as this step is not must for the user, but a choice.

sterobin commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-external-application-rest.adoc:
```diff
@@ -0,0 +1,15 @@
+To migrate a DMN project to {KOGITO} deployment, first you can migrate external applications that are interacting with decision services on the {KIE_SERVER}, using {KIE_SERVER} REST endpoints that are specific to DMN models.
+
+.Procedure
+. To migrate your external application to a REST endpoint specific to a DMN model, perform the following steps:
```
Delete this parent step and make the others standard steps. Not clear on why it's structured otherwise.

sterobin commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-external-application-rest.adoc:
```diff
@@ -0,0 +1,15 @@
+
+.Procedure
+. To migrate your external application to a REST endpoint specific to a DMN model, perform the following steps:
+.. If you are using REST endpoints in your external application, retrieve the Swagger or OAS specification file of the KJAR using `GET /server/containers/{containerId}/dmn/openapi.json (|.yaml)` endpoint.
```
"using **the** ... endpoint"

sterobin commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-external-application-rest.adoc:
```diff
@@ -0,0 +1,15 @@
+
+.. In your external application, select the Java or JDK library to interact with the decision services. You can interact with the decision services using the REST endpoint for the specific KJAR.
+
+NOTE: The {KIE_SERVER} Client Java API is not supported in migration to {KOGITO} deployment.
```
"in **the** migration to **a**"

sterobin commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-dmn-model-kjar.adoc:
```diff
@@ -0,0 +1,35 @@
+[id="proc-kogito-migrate-dmn-model-kjar_{context}"]
+= Migrating DMN model KJAR to {KOGITO} microservice
```
"Migrating **a** DMN model KJAR to **a** {KOGITO} microservice"

So @hmanwani-rh I'll try to make this the last time I point out the missing articles just to spare you :) Please do have a look through all headings and really re-review the whole doc for all the missing articles. I'm not clear on why the language is that way, but it's non-standard and not advised per IBM and most tech comm style manuals. Let me know if I've misunderstood anything though.

sterobin commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-dmn-model-kjar.adoc:
```diff
@@ -0,0 +1,35 @@
++
+The Maven project creates Kogito artifacts.
+
+. Copy the DMN models from `src/main/resources` folder of the KJAR to `src/main/resources` folder of Kogito artifact.
```
"from **the** ... folder" and "to **the** ... folder" "of **the**... artifact"  (sorry, I know I said I'd stop :) )

sterobin commented on doc-content/enterprise-only/getting-started-kogito/ref-migrate-kjar-dmn-example.adoc:
```diff
@@ -0,0 +1,89 @@
+[id="ref-migrate-kjar-dmn-example_{context}"]
+= Example of migrating DMN model KJAR to {KOGITO} microservice
+
+Following is an example of migrating a DMN model KJAR to {KOGITO} microservice:
```
I see several instances of "Following is an example" Technically this is syntactically correct, but definitely not common and non-standard. I'd suggest at least "The following is ...", which IBM still says to avoid in favor of "The following example illustrates/shows/whatever" for concreteness. But either way, I suggest we be sure to avoid "Following is" because most will view that is incorrect or too informal for enterprise-level docs. Suggest so throughout.

sterobin commented on doc-content/enterprise-only/getting-started-kogito/ref-example-bind-external-app-kogito.adoc:
```diff
@@ -0,0 +1,15 @@
+
+After deploying the {KOGITO} microservice, you need to bind your external application to the {KOGITO} microservice deployment.
+
+Binding your external application includes re-routing the external application and binding the application to a new base URL of the server that is associated with the {KOGITO} application. For example:
```
I see several instances of "For example:" throughout. Just a reminder that IBM says to use complete sentences before colons and uses "For example:" as an instance to avoid. Instead, suggests something like "..., as shown in the following example:" or "For example, the following endpoint is/demonstrates/whatever ..." or even just "The following endpoint example illustrates ..." Something. Again a suggestion throughout anytime that appears in the doc.

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-pmml-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,46 @@
+[id="con-migration-pmml-strategy_{context}"]
+== Migration strategy
+
+The migration strategy of migrating PMML project to {KOGITO} deployment, includes the following steps:
```
Again no comma, just "deployment includes ..."

sterobin commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-dmn-model-kjar.adoc:
```diff
@@ -0,0 +1,35 @@
+mvn clean install
+----
+
+After running the {KOGITO} application, you can retrieve the Swagger or OAS specification file. The Swagger or OAS specifications provide the same information as the REST endpoint along with some implementation details, such as:
```
Per IBM should be "such as the following information:" Again because of IBM's preference to keep the sentence complete before the colon. This comes up a couple times, so applies elsewhere too. Search all.

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-pmml-to-kogito-overview.adoc:
```diff
@@ -0,0 +1,46 @@
+
+The following table describes the major changes and features that affect migration from {KIE_SERVER} API and KJAR to {KOGITO} deployments:
+
+.Migration to PMML considerations
```
Again suggest "PMML migration considerations"

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-drl-to-kogito-loan-overview.adoc:
```diff
@@ -0,0 +1,58 @@
+
+The following table describes the major changes and features that affect migration from {KIE_SERVER} API and KJAR to {KOGITO} deployments:
+
+.Migration to DMN considerations
```
Typo in label, but also again suggest "DRL migration considerations"

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-drl-to-kogito-loan-overview.adoc:
```diff
@@ -0,0 +1,58 @@
+
+Using legacy API in {KOGITO}
+Migrating to {KOGITO} rule units
+
```
This streamlines a bit imo, and gets rid of the empty "there is/are" construction that IBM says to avoid, and avoids the rendering problem that I'm currently seeing in your output:

```
In {PRODUCT}, you can migrate a rule evaluation to a {KOGITO} deployment in the following two ways:

Using legacy API in {KOGITO}::
In {KOGITO}, ...

Migrating to {KOGITO} rule units::
Migrating to ...
```

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-drl-to-kogito-loan-overview.adoc:
```diff
@@ -0,0 +1,58 @@
+Migrating to {KOGITO} rule units
+
+Using legacy API in {KOGITO}::
+In {KOGITO}, the `kogito-legacy-api` module makes the legacy API of {RHPAM} available, therefore the DRL files remain unchanged. This approach of migrating rules evaluation requires minimal changes and enables you to use major {QUARKUS} features, such as hot reload and native image creation.
```
The RHPAM attribute isn't rendering, FYI.

Also suggest "available so that the DRL files ..." If you use therefore, then the sentence structure as to be changed to a semicolone "available; therefore, the .." or two sentences.

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-drl-to-kogito-loan-overview.adoc:
```diff
@@ -0,0 +1,58 @@
+Using legacy API in {KOGITO}::
+In {KOGITO}, the `kogito-legacy-api` module makes the legacy API of {RHPAM} available, therefore the DRL files remain unchanged. This approach of migrating rules evaluation requires minimal changes and enables you to use major {QUARKUS} features, such as hot reload and native image creation.
+
+Migrating to {KOGITO} rule units::
```
When you have paragraph breaks in block units like this, you need to enclose the segment with `+ --` starting and `--` ending to maintain the indentation. Right now it's off in the output.

So like this:

```
Migrating to {KOGITO} rule units::
+
--
Migrating to {KOGITO} rule units ...

A rule unit ...

Overall, ...
--
```

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-drl-to-kogito-loan-overview.adoc:
```diff
@@ -0,0 +1,58 @@
+Migrating to {KOGITO} rule units include the programming model of {KOGITO}, which is based on the concept of rule units.
+
+A rule unit in {KOGITO} includes both a set of rules and the facts, against which the rules are matched. Rule units in {KOGITO} also come with data sources. A rule unit data source is a source of the data processed by a given rule unit and represents the entry point, which is used to evaluate the rule unit. There are two types of data sources:
+
```
Suggest the following. Typically best for a lead-in to a list to be a dedicated light/paragraph break. Also avoid the empty "there are".
```
... evaluate the rule unit.

Rule units use two types of data sources:

* `DataStream`: ....
```

sterobin commented on doc-content/enterprise-only/getting-started-kogito/con-migrate-drl-to-kogito-loan-overview.adoc:
```diff
@@ -0,0 +1,58 @@
+* `DataStream`: This is an append-only data source and the facts added into the `DataStream` cannot be updated or removed.
+* `DataStore`: This data source is for modifiable data. You can update or remove an object using the `FactHandle` that is returned when the object is added into the `DataStore`.
+
+Overall, a rule unit contains two related parts, including the definition of the fact to be evaluated and the set of rules evaluating the facts.
```
Suggest "two parts: The definition of ..." Reason is that "including" implies that there are X parts, include these but also some others. When it's literally just the two things, best to just say so explicitly.

sterobin commented on doc-content/enterprise-only/getting-started-kogito/ref-kogito-migrate-example-loan-application-project.adoc:
```diff
@@ -0,0 +1,70 @@
+= Example loan application project
+
+[role="_abstract"]
+In the following sections, loan application project is used as an example to migrate a DRL project to {KOGITO} deployments. The domain model of the loan application project is made of two classes, including `LoanApplication` class and `Applicant` class as follows:
```
Sorry, but this one is striking: "sections, **a** loan application ..."

Also suggest "made of two classes, the ...  class and the ... class:"  Here you can go with the comma instead of the colon because it's then followed by a colon for the example. Also avoids the "as follows" that IBM says to avoid.

sterobin commented on doc-content/enterprise-only/getting-started-kogito/ref-kogito-migrate-example-loan-application-project.adoc:
```diff
@@ -0,0 +1,70 @@
+}
+----
+
+The rules set is created using business decisions to approve or reject an application, along with the last rule of collecting all the approved applications in a list.
```
"The rule set" singular, as you then say in the next snippet.

sterobin commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-drl-legacy-api.adoc:
```diff
@@ -0,0 +1,61 @@
+[id="proc-kogito-migrate-drl-legacy-api_{context}"]
+= Migrating rules units to a {KOGITO} microservice using legacy API
```
It's "rule units". Same elsewhere in a ref. You'll see in a search.

sterobin commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-drl-legacy-api.adoc:
```diff
@@ -0,0 +1,61 @@
+After exposing the rule units with a REST endpoint, you can migrate the rule units to a {KOGITO} microservice using legacy API.
+
+.Procedure
+. Add the following dependencies to the project `pom.xml` file to enable the use {QUARKUS} and legacy API
```
Missing punctuation, I assume colon "API:"

sterobin commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-drl-implement-ruleunits.adoc:
```diff
@@ -0,0 +1,152 @@
+* `DataStream`: This is an append-only data source. In `DataStream`, subscribers receive new and past messages, stream can be hot or cold in the reactive streams. Also, the facts added into the `DataStream` cannot be updated or removed.
+* `DataStore`: This data source is for modifiable data. You can update or remove an object using the `FactHandle` that is returned when the object is added into the `DataStore`.
+
+Overall, a rule unit contains two related parts, including the definition of the fact to be evaluated and the set of rules evaluating the facts.
```
I see you've reused this content from before, so similar suggestions.

So this again:

```
... evaluate the rule unit.

Rule units use two types of data sources:

* `DataStream`: ....
```

And suggest "two parts: The definition of ..."

sterobin commented on doc-content/enterprise-only/getting-started-kogito/proc-kogito-migrate-drl-implement-ruleunits.adoc:
```diff
@@ -0,0 +1,152 @@
+Overall, a rule unit contains two related parts, including the definition of the fact to be evaluated and the set of rules evaluating the facts.
+
+.Procedure
+. Implement both parts of a rule unit, including fact definition and DRL file using the following steps:
```
Again I don't see the need for this parent step, as the two steps that follow are the only two in the proc anyway. Suggest deleting this parent step and just going with the actual steps that follow.

sterobin reviewed (commented):
@hmanwani-rh It's so great that Kogito is to this stage! Very exciting. Don't be alarmed, but I did leave quite a lot of comments, so be sure to expand the "Hidden conversations" in the PR. The content overall does still feel a bit rough, with lots of little things that I've pointed out throughout. My overall concern is that there are a lot of dropped articles, which makes the language throughout pretty non-standard and not quite enterprise-level quality. I stopped tracking them after awhile and have left it to you to go through and tidy up all missing articles etc.

I'm so sorry to leave such a heavy review :( I think it'll go fast, just a fair amount of tidying up still needed imo.

hmanwani-rh commented:
> @hmanwani-rh It's so great that Kogito is to this stage! Very exciting. Don't be alarmed, but I did leave quite a lot of comments, so be sure to expand the "Hidden conversations" in the PR. The content overall does still feel a bit rough, with lots of little things that I've pointed out throughout. My overall concern is that there are a lot of dropped articles, which makes the language throughout pretty non-standard and not quite enterprise-level quality. I stopped tracking them after awhile and have left it to you to go through and tidy up all missing articles etc.
> 
> I'm so sorry to leave such a heavy review :( I think it'll go fast, just a fair amount of tidying up still needed imo.

@sterobin Thank you for your thorough review. 
I've updated the files as suggested and also fixed the articles.
